File: C++/A/366A-dimaandguards.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> can_pass_guardpost(int budget, array<int, 4> prices)
{
    vector<pair<int, int>> combinations = {{prices[0], prices[2]}, {prices[0], prices[3]}, {prices[1], prices[2]}, {prices[1], prices[3]}};
    for (const auto &item : combinations)
        if (item.first + item.second <= budget)
            return {item.first, budget - item.first};

    return {-1};
}

vector<int> sneak_in_inna2(int budget, array<array<int, 4>, 4> guardposts)
{
    for (int i : {0, 1, 2, 3})
    {
        vector<int> result = can_pass_guardpost(budget, guardposts[i]);
        if (result[0] != -1)
            return {i + 1, result[0], result[1]};
    }

    return {-1};
}

int main()
{
    int budget;
    cin >> budget;

    array<array<int, 4>, 4> guardposts;
    for (int i : {0, 1, 2, 3})
        for (int j : {0, 1, 2, 3})
            cin >> guardposts[i][j];

    vector<int> result2 = sneak_in_inna2(budget, guardposts);
    if (result2[0] != -1)
        for (int value : result2)
            cout << value << " ";
    else
        cout << -1;
}

File: Python/366A-dimaandguards.py
def can_pass_guardpost(budget, prices):
    combinations = [(prices[0], prices[2]), (prices[0], prices[3]),
                    (prices[1], prices[2]), (prices[1], prices[3])]
    for item in combinations:
        if sum(item) <= budget:
            return [item[0], budget - item[0]]
    return -1


def sneak_in_inna(budget, guardposts):
    for i, guardpost in enumerate(guardposts):
        result = can_pass_guardpost(budget, guardpost)
        if result != -1:
            return [i + 1] + result
    return -1


budget = int(input())
guardposts = []
for _ in range(4):
    guardposts.append(list(map(int, input().split())))

result = sneak_in_inna(budget, guardposts)

print(*result) if result != -1 else print(-1)
